### ISSUE TITLE
Dataset from generator is far slower than from tensor slices, anything I can improve?

### ISSUE THREAD

**Issue by hkvision** [tensorflow/tensorflow#56258]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Performance

### Source

binary

### Tensorflow Version

tf 1.15.0

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 18.04

### Python version

3.7

### Current Behaviour?

```shell
When I use `tf.data.Dataset.from_tensor_slices` and `tf.data.Dataset.from_generator` to create the same dataset, I observe when fetching batches, from_generator is much slower than from_tensor_slices (more than 10x slower).
Also the performance of from_generator greatly drops when the number of yield elements increases. (See the details below).

I am using tf1.15, but I tested tf2.8, seems the performance of generator is even worse...

Wondering is it reasonable that Dataset from generator is far slower than from tensor slices? And want some help on how to improve the performance using the generator. If I want to have a dataset with 10 elements for each records, how can I get better performance?

Thanks so much in advance!
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np
import time

tf.compat.v1.enable_eager_execution(
    config=None, device_policy=None, execution_mode=None
)

size = 100000
data = np.random.rand(size)

def get_one():
    i = 0
    while i < size:
        yield tuple([data[i]]*10)
        i += 1

# dataset = tf.data.Dataset.from_generator(get_one, output_types=tuple([tf.float32]*10))
dataset = tf.data.Dataset.from_tensor_slices(tuple([data]*10))
dataset = dataset.batch(512)

i = 0
total_time = 0
start = time.time()
for sample in dataset:
    # Performing a training step
    end = time.time()
    used = end - start
    total_time += used
    print("Get batch time: ", used)
    i += 1
    start = time.time()
print("Average get batch time: ", total_time / i)
```

### Relevant log output

When using from_tensor_slices, average get batch time:  0.001715s
When using from_generator, average get batch time:  0.10747s [0.14395s for tf2.8]

If I change the generator function as follows and re-run the above program:
```shell
def get_one():
    i = 0
    while i < size:
        for j in range(10):
            res = data[i]
        yield res
        i += 1
```
Namely if I get the value 10 times in the generator, but only yield one element, the performance is much faster compared with yield 10 features: average get batch time:  0.02014s, even though it is still much slower than from_tensor_slices. Why is that? I suppose the computation is the same, the only difference is how many elements I output? (BTW if I use tf2.8, in this case, it is 0.3935s, unreasonable... don't know why...)

**Comment by tilakrayal** [comment]
@hkvision ,
I was able to execute the code without any issues in tensorflow v2.8 and also the average batch time for both are similar. 


```
Average get batch time:  0.0023131625992911203
Average get batch time:  0.0024135222240370146
```


Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/dc3cf17e0831e78f861ceef966297f59/56258.ipynb). Thank you!

**Comment by hkvision** [comment]
Hi @tilakrayal 

Thanks for your reply!

But in your notebook, the two code segments are both based on from_tensor_slices, but not from_generator...😂
I changed to from_generator:
```
dataset = tf.data.Dataset.from_generator(get_one, output_types=tf.float32)
# dataset = tf.data.Dataset.from_tensor_slices(tuple([data]*10))
```
and I get the result: Average get batch time:  0.3002346827059376, more than 100x slower... could you help verify this and give some help on how to improve?

**Comment by tilakrayal** [comment]
@hkvision,
I was facing different error while trying to execute the mentioned code. Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/477ac53b13c581843b3da03b4b6745b4/untitled357.ipynb).

**Comment by hkvision** [comment]
Changing
```
dataset = tf.data.Dataset.from_generator(get_one, output_types=tuple([tf.float32]*10))
```
to 
```
dataset = tf.data.Dataset.from_generator(get_one, output_types=tf.float32)
```
should work.

Or change the generator function to return 10 results.

Take a further look?

**Comment by tilakrayal** [comment]
@sachinprasadhs ,
I was able to reproduce the issue in tensorflow [v2.8](https://colab.research.google.com/gist/tilakrayal/b64d35413de7538daaece7ff38db2fbe/2-8_56258.ipynb), [v2.9](https://colab.research.google.com/gist/tilakrayal/60802dec090d00c9df224fa3e227ef08/2_9_56258.ipynb) and [nightly](https://colab.research.google.com/gist/tilakrayal/8f19eadcd6ae01d357db3ca2f730bf80/2_10_56258.ipynb). Kindly find the attached gist.

**Comment by sachinprasadhs** [comment]
Hi, arguments like `output_shapes, output_types` will be deprecated in future version, can you check with `output_signature` and let us know if you find similar behavior. Thanks!

**Comment by hkvision** [comment]
I will check `output_signature`...
@google-ml-butler Please wait and not close this issue at this moment :)

**Comment by sachinprasadhs** [comment]
@hkvision , Any update from the above comment, did you find any difference in performance, also try using tf-nightly and latest Tensorflow 2.9.1 version.

**Comment by hkvision** [comment]
I will test and give you the result as soon as possible.

**Comment by aaudiber** [comment]
`Dataset.from_generator` is generally not as performant as other ways of using tf.data, since it executes the input generation logic in Python instead of C++. In particular, the generator will always run single-threaded, while computation expressed with other tf.data transformations can be run with many threads in parallel. If performance is critical, prefer other APIs like `Dataset.from_tensor_slices`

**Comment by hkvision** [comment]
@aaudiber Okay, I understand, thank you for your reply!

**Comment by sachinprasadhs** [comment]
@hkvision , If your concern is addressed, could you please close the issue. Thanks!

**Comment by hkvision** [comment]
Sure. Thanks for the help.

**Comment by ofiryaish** [comment]
> `Dataset.from_generator` is generally not as performant as other ways of using tf.data, since it executes the input generation logic in Python instead of C++. In particular, the generator will always run single-threaded, while computation expressed with other tf.data transformations can be run with many threads in parallel. If performance is critical, prefer other APIs like `Dataset.from_tensor_slices`

@aaudiber @tensorflowbutler @sachinprasadhs 
The problem with`Dataset.from_tensor_slices` is that it is unsuitable for big datasets, while `Dataset.from_generator` will work.
There are many technical problems with `from_generator`, and the main one, in my opinion, is that loading time is extremely slow compared to `Dataset.from_tensor_slices`. in my case, using generators slows the training in x3. This is a big problem!

Therefore, I think this issue should be open and treated as for big datasets, `Dataset.from_generator` should provide an easy solution.

**Comment by aaudiber** [comment]
If you care about performance with large datasets try using a source dataset appropriate for your data format, such as `tf.data.TFRecordDataset`. Trying to make Python input processing with `from_generator` as performant as C++ processing is going to be an uphill battle.

**Comment by ofiryaish** [comment]
> If you care about performance with large datasets try using a source dataset appropriate for your data format, such as `tf.data.TFRecordDataset`. Trying to make Python input processing with `from_generator` as performant as C++ processing is going to be an uphill battle.

@aaudiber 

I used tf.data.TFRecordDataset, however, the writing-to-file process is not scaling up well. I tried to write 400k samples into a TFRecord file, which took me a couple of minutes. However, when I wrote 40 million samples, It took me around 10 hours. I think that there is a problem with the way tf.io.TFRecordWriter is writing. We cannot control the buffer which determines when to perform disk access - meaning that I do not know how to write in bulks of specific size. Note that this is my assumption for not being able to scale up well.

**Comment by aaudiber** [comment]
It might help to write to multiple files in parallel. If you use Dataset.save/Dataset.load, it will automatically parallelize the writing for you.

**Comment by ofiryaish** [comment]
> It might help to write to multiple files in parallel. If you use Dataset.save/Dataset.load, it will automatically parallelize the writing for you.

@aaudiber 

This is indeed one option; however, it seems that even without parallelizing the process, the execution time should be linear as a function of the number of samples, but it is not linear. Therefore, I assume that the bottleneck is the number of disk access that costs too much. If that is the problem, I think that allowing to write to disk in bulks will improve the execution time significantly (and control the bulk/buffer size). It might be that I'm wrong, and it already writes in bulks, but I'm not familiar with the interior implementation.

Thank you

**Comment by ofiryaish** [comment]
@aaudiber

I did a little test. This is part of my code (I know that each time I overwrite the file repeatedly). I followed the file size during the process and found that the increase in size in each update is very small (probably less than MB). This confirms what I thought, and the bottleneck is in the number of times `tf.data.experimental.TFRecordWriter` writes. This is also the case with `tf.io.TFRecordWriter`.

```
for i in range(0, len(self.y), self.write_buffer_size):
    def serialized_gen():
        for j in range(i, min(len(self.y), i + self.write_buffer_size)):
            x_dict = {"x_{}".format(k): self._bytes_feature(
                self._serialize_array(self.x[k][j])) for k in range(len(self.x))}
            y_dict = {"y": self._bytes_feature(self._serialize_array(self.y[j]))}
            sample_weight_dict = {} if self.sample_weight is None else \
                {"sample_weight": self._bytes_feature(self._serialize_array(self.sample_weight[j]))}
            feature = {**x_dict, **y_dict, **sample_weight_dict}
            yield tf.train.Example(features=tf.train.Features(feature=feature)).SerializeToString()

    serialized_features_dataset = tf.data.Dataset.from_generator(
        serialized_gen, output_types=tf.string, output_shapes=())
    writer = tf.data.experimental.TFRecordWriter(self.dataset_path)
    writer.write(serialized_features_dataset)
```

**Comment by ofiryaish** [comment]
> It might help to write to multiple files in parallel. If you use Dataset.save/Dataset.load, it will automatically parallelize the writing for you.

@aaudiber 

I got additional questions regarding this answer. I tried to divide the dataset and parallelize the writing using multiprocessing (each worker writes with `tf.io.TFRecordWriter`), but it did not work for me. Would you happen to have a template for how to do that?

Regarding the `tf.data.Dataset.save/load`, Does it perform the same as TFrecords? The advantage of using TFrecords is being able to load part of the data each time and, therefore, not getting memory allocation problems.
